Refactor to encapsulate systems

spikeheap opened the pull request:
## What does this pull request do?

This PR refactors CloudPlatform, NOMIS and PrisonerContentHub to be singleton objects where the models are defined in the constructor, and relationships and views are defined in separate steps.

This is a riff on the [Page Object Model](https://artoftesting.com/pageObjectModel) for end-to-end testing.

Following this pattern, to add a new system you would need to make 2 changes:

1. Add the system, following the pattern used in NOMIS
2. Add the new system to `[prison/probation]/workspace.kt`

## What is the intent behind these changes?

I've created it to experiment with ways we might ease the creation of new systems, and keep related components close to each other (e.g. views of a system alongside the models), rather than sock-drawer separation (models, relationships, views).

By separating the relationship and model steps, we can avoid circular dependency/first defined problems. All models are present at the point the relationships are defined.

This approach also demonstrates having system objects to make interacting with the domain easier, e.g.:

```
val nomis = NOMIS.getInstance(model)
```

At the very least, this would need a parent class or interface, and there's probably a much easier way to achieve the same goal!

Also, `SingletonHolder` is copied almost verbatim from [here](https://medium.com/@BladeCoder/kotlin-singletons-with-argument-194ef06edd9e). It melted my brain trying to figure out what it was doing (or how to add additional arguments to the singleton constructor), but it worked to create a singleton.

spikeheap commented:
Thanks for the multiple reviews @sldblog, I think this _might_ be good to go now

sldblog commented on src/main/kotlin/prison/model.kt:
```diff
@@ -9,15 +9,10 @@ fun prisonModel(model: Model) {
   model.setImpliedRelationshipsStrategy(
       CreateImpliedRelationshipsUnlessAnyRelationshipExistsStrategy())
 
-  val nomis = NOMIS(model)
```
What do you think about initialising all model objects here instead of `workspace.kt`? I'd be confused if I'd see Nomis on the model without seeing it at the same level as others.

Having it in the same file would make it more prominent that there's a transition in progress.

spikeheap replied:
Hi @sldblog I'd rather refactor these models out of here in a subsequent PR. Having the model -> relationship -> view steps in `workspace.kt` makes it more cohesive, and allows us to iterate over the classes rather than defining them in multiple places. 

I don't mind doing the legwork to refactor the other classes out after this is merged. We could also:
- Add a comment into this file to say not to use it, or
- Move the model definitions up into `workspace.kt` and remove this file. It's only a few lines, so should be quite obvious there.

I've added a comment to this file to make it clearer and hopefully reduce confusion.

sldblog replied:
👍 Sounds good. Hopefully, we can remove the comments soon (I don't like comments :neckbeard:)

sldblog commented on src/main/kotlin/prison/workspace.kt:
```diff
@@ -12,10 +14,41 @@ fun prisonWorkspace(): Workspace {
+    PrisonerContentHub
+  )
+
+  // We start by defining the deployment nodes. Containers are often associated with deployment nodes
+  // during model definition, so we front-load this task to make everything available for that step.
+  CloudPlatform.defineDeploymentNodes(workspace.model)
```
What do you think about removing the comments? They have a nasty habit of getting out of date quickly.

spikeheap replied:
Comments feel useful for explaining the intent of the code, and here it feels like they're useful to convey the importance of ordering. 

I would _hope_ that if someone changes the order or introduces other dependencies, either they'll update these comments or we'd catch it in a pull request, but I'm not sure I agree that it's better _not_ to have them in case they go out of date 🤷‍♂️.

sldblog replied:
Maybe it's possible to make that implicit ordering clearer with the names 🤔

sldblog commented on src/main/kotlin/shared/cloudPlatform.kt:
```diff
@@ -1,13 +1,29 @@
+  
+
+  companion object {
+    lateinit var cloudPlatform: DeploymentNode
```
Do we need to expose this top-level node?

spikeheap replied:
Removed in 97145bc00b7f394065b1f58b845aacfe4c6ec0c1

sldblog commented on src/main/kotlin/prison/nomis.kt:
```diff
@@ -4,83 +4,97 @@ import com.structurizr.model.Location
+      val cloudPlatform = model.getDeploymentNodeWithName("Cloud Platform")
+      val sqs = cloudPlatform.getDeploymentNodeWithName("SQS")
+      val sns = cloudPlatform.getDeploymentNodeWithName("SNS")
+      val kubernetes = cloudPlatform.getDeploymentNodeWithName("Kubernetes")
```
Would it be possible to replace name-based lookups with `CloudPlatform.kubernetes`?

spikeheap replied:
:+1: I've updated this now

sldblog commented:
> This is not meant to be merged.

Do we need to change the pull request description?

spikeheap commented:
Hi @sldblog, I think I've addressed all the points. If we agree with the principle of this change, I'd like us to merge it and push further improvements/changes into subsequent PRs instead of this one, so I can break free from fixing merge conflicts every couple of days :D.
